widgets/nuxeo-uploader-behavior.html

```diff
@@ -17,55 +17,244 @@
+     *
+     * @return {Boolean} true if the provider reports progress, false otherwise
+     **/
+    DefaultUploadProvider.prototype.hasProgress = function() {
+      return false;
+    };
+
+    var _registry = { 'default': DefaultUploadProvider };
+    var _defaultProvider = 'default';
+
+    /**
+     * @polymerBehavior Nuxeo.UploaderBehavior
+     */
+    Nuxeo.UploaderBehavior = {
+
+      properties: {
+        /**
+         * Accepted file extensions or mime types (comma separated values).
+         */
+        accept: String,
+
+        /**
+         * This flag determines whether the file should be immediately uploaded or not.
+         */
+        immediate: {
+          type: Boolean,
+          value: true
+        },
+
+        /**
+         * Current batch id.
+         */
+        batchId: String,
+
+        /**
+         * List of files in the current batch.
+         */
+        files: {
+          type: Array,
+          value: []
+        },
+
+        /**
+         * Flag that indicates if an upload is in progress.
+         */
+        uploading: {
+          type: Boolean,
+          value: false
+        },
+
+        /**
+         * Allow multiple files to be added to the same batch.
+         */
+        batchAppend: {
+          type: Boolean,
+          value: false
+        },
+
+        /**
+         * Current upload provider being used. If `undefined`, the `defaultProvider` will be used.
+         */
+        provider: {
+          type: String,
+          notify: true,
+          reflectToAttribute: true,
+          observer: '_reloadProvider'
+        },
+
+        _provider: {
+          type: String,
+          value: function() {
+            return _defaultProvider;
+          }
+        }
+      },
+
+      observers: ['_initProvider(_provider, connection, accept, batchAppend)'],
+
+      _initProvider: function() {
+        if (this._provider) {
+          this._instance = new _registry[this._provider](this.connection, this.accept, this.batchAppend);
+        }
+      },
+
+      _reloadProvider: function() {
+        if (this.provider && this.provider !== this._provider) {
+          this.set('_provider', this.provider);
+        } else if (!this.provider) {
+          this.set('_provider', _defaultProvider);
+        }
+      },
+
+      created: function() {
+        this.defaultProviderChangedHandler = this._reloadProvider.bind(this);
+        document.addEventListener('defaultUploadProviderChanged', this.defaultProviderChangedHandler);
       },
 
       /**
-       * Current batch id.
+       * Registers a new upload provider.
        */
-      batchId: String,
+      registerProvider: function(name, provider) {
+        _registry[name] = provider;
+      },
 
       /**
-       * List of files in the current batch.
+       * Returns all registered upload providers.
        */
-      files: {
-        type: Array,
-        value: []
+      getProviders: function() {
```

Comment:
Do we want to expose this?

Reply:
For now we're only using it for debugging.